components/tidb_query/src/codec/convert.rs

```diff
@@ -443,31 +443,49 @@ impl ToInt for Json {
     #[inline]
     fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
         // Casts json to int has different behavior in TiDB/MySQL when the json
-        // value is a `Json::Double` and we will keep compatible with TiDB
+        // value is a `Json::from_f64` and we will keep compatible with TiDB
         // **Note**: select cast(cast('4.5' as json) as signed)
         // TiDB:  5
         // MySQL: 4
-        let val = match *self {
-            Json::Object(_) | Json::Array(_) | Json::None | Json::Boolean(false) => Ok(0),
-            Json::Boolean(true) => Ok(1),
-            Json::I64(d) => Ok(d),
-            Json::U64(d) => Ok(d as i64),
-            Json::Double(d) => d.to_int(ctx, tp),
-            Json::String(ref s) => s.as_bytes().to_int(ctx, tp),
+        let val = match self.as_ref().get_type() {
+            JsonType::Object | JsonType::Array => Ok(0),
+            JsonType::Literal => Ok(match self.as_ref().get_literal() {
+                Some(b) => {
+                    if b {
```

Comment:
How about `b as i64`

Reply:
a more rusty code is `self.as_ref().get_literal().map(|x| x as i64).unwrap_or(0)`, but i'm not sure the readability is better.